// fnbr/dist/src/util/AsyncLock.js
"use strict";
/*
  Copyright © 2020 The Sapphire Community and its contributors

  Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the “Software”), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Represents an asynchronous task lock used for pending token refreshing or party changes
 * @private
 */
class AsyncLock {
    constructor() {
        this.lockPromise = undefined;
    }
    /**
     * Whether this lock is active
     */
    get isLocked() {
        return !!this.lockPromise;
    }
    /**
     * Returns a promise that will resolve once the lock is released
     */
    wait() {
        var _a;
        return ((_a = this.lockPromise) === null || _a === void 0 ? void 0 : _a.promise) || Promise.resolve();
    }
    /**
     * Activates this lock
     */
    lock() {
        let resolve;
        const promise = new Promise((res) => {
            resolve = res;
        });
        this.lockPromise = { promise, resolve: resolve };
    }
    /**
     * Deactivates this lock
     */
    unlock() {
        var _a;
        (_a = this.lockPromise) === null || _a === void 0 ? void 0 : _a.resolve();
        this.lockPromise = undefined;
    }
}
exports.default = AsyncLock;
